Add "batching rule" for torch.sum(tensor, dims)

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38723 Add "batching rule" for torch.sum(tensor, dims)**
* #38798 Fixup: rename BatchedTensorKey to Batched

On the way to moving vmap prototype into tree, this is the
implementation of the sum batching rule and some utility functions.
After implementing this and another batching rule (probably for
at::mul), we'll be ready to for the initial vmap Python
API.

Re: testing: There's a bit of a chicken-and-egg problem here; I'd like to test
that the batching rules work for vmap, but the easiest way to do that is using
the (non-existent) vmap python API. This PR includes some minimal testing for
this batching rule.

Future PRs (for more context):
- [YOU ARE HERE] Add torch.sum(tensor, dims) batching rule + some helper functions
- Add torch.mul batching rule + some more helper functions
- Add vmap python API
- Add "slow" for-loop fallbacks for out-of-place functions via
dispatcher fallback mechanism.
- Add batching rules for "view" functions
- Add "slow" for-loop fallbacks for in-place functions
- And more

Test Plan:
- `./build/bin/vmap_test`

zou3519 committed:
Add "batching rule" for torch.sum(tensor, dims)

On the way to moving vmap prototype into tree, this is the
implementation of the sum batching rule and some utility functions.
After implementing this and another batching rule (probably for
at::mul), we'll be ready to for the initial vmap Python
API.

Batching rules and utility functions act on "unpacked BatchedTensors".
This is to avoid tensor allocation overhead (one alternative is having
these act on BatchedTensors). Unfortunately this leads to more verbose
and confusing code. I've considered adding a special
"BatchedTensorData" struct that only holds a Tensor and BatchDims and
has helper methods defined on it but I'm not sure if that is better.

Re: testing: There's a bit of a chicken-and-egg problem here; I'd like to test
that the batching rules work for vmap, but the easiest way to do that is using
the (non-existent) vmap python API. This PR includes some minimal testing for
this batching rule.

Future PRs (for more context):
- [YOU ARE HERE] Add torch.sum(tensor, dims) batching rule + some helper functions
- Add torch.mul batching rule + some more helper functions
- Add vmap python API
- Add "slow" for-loop fallbacks for out-of-place functions via
dispatcher fallback mechanism.
- Add batching rules for "view" functions
- Add "slow" for-loop fallbacks for in-place functions
- And more

Test Plan:
- `./build/bin/vmap_test`

[ghstack-poisoned]

ezyang commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -0,0 +1,34 @@
+  TORCH_CHECK(false, "NYI: Calling ", op.schema().name(), " inside of vmap");
+}
+
+TORCH_LIBRARY_IMPL(_, BatchedTensorKey, m) {
```
Oops, I missed this in the earlier PR. Please call this just `Batched`, not `BatchedTensorKey`!

zou3519 replied:
Will do, that makes sense

ezyang commented on aten/src/ATen/BatchingUtils.cpp:
```diff
@@ -0,0 +1,55 @@
+}
+
+// NB: There is an invariant that BatchDims are stored in increasing
+// `level` order.
```
This should just be a `See Note [...]` to https://github.com/pytorch/pytorch/pull/38424/files#diff-254a034cc0b96a370756c251b786e92cR81

ezyang commented on aten/src/ATen/BatchingUtils.h:
```diff
@@ -0,0 +1,39 @@
+/*
+ * Utility functions used to implement batching rules.
+ *
+ * NB: All of these do NOT accept Tensors backed by BatchedTensor, unless
```
Add some internal asserts to test for this invariant?

zou3519 replied:
Done

ezyang commented on aten/src/ATen/BatchingUtils.h:
```diff
@@ -0,0 +1,39 @@
+// this function returns the underlying Tensor and BatchDims.
+// If the input is a Tensor backed with regular TensorImpl, then
+// this function returns the tensor and empty BatchDims.
+TORCH_API std::pair<Tensor, BatchDimsRef> unpackBatched(const Tensor& self);
```
It's weird that this returns an owning reference to the tensor, but a reference to batch dims. If the validity of the output argument is going to rely on the input tensor staying live, might as well just return references for both?

zou3519 committed:
Update on "Add "batching rule" for torch.sum(tensor, dims)"


On the way to moving vmap prototype into tree, this is the
implementation of the sum batching rule and some utility functions.
After implementing this and another batching rule (probably for
at::mul), we'll be ready to for the initial vmap Python
API.

Batching rules and utility functions act on "unpacked BatchedTensors".
This is to avoid tensor allocation overhead (one alternative is having
these act on BatchedTensors). Unfortunately this leads to more verbose
and confusing code. I've considered adding a special
"BatchedTensorData" struct that only holds a Tensor and BatchDims and
has helper methods defined on it but I'm not sure if that is better.

Re: testing: There's a bit of a chicken-and-egg problem here; I'd like to test
that the batching rules work for vmap, but the easiest way to do that is using
the (non-existent) vmap python API. This PR includes some minimal testing for
this batching rule.

Future PRs (for more context):
- [YOU ARE HERE] Add torch.sum(tensor, dims) batching rule + some helper functions
- Add torch.mul batching rule + some more helper functions
- Add vmap python API
- Add "slow" for-loop fallbacks for out-of-place functions via
dispatcher fallback mechanism.
- Add batching rules for "view" functions
- Add "slow" for-loop fallbacks for in-place functions
- And more

Test Plan:
- `./build/bin/vmap_test`

[ghstack-poisoned]

ezyang commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -0,0 +1,34 @@
+  BatchDims result_bdims;
+
+  std::tie(self_, self_bdims) = unpackBatched(self);
+  std::tie(result_, result_bdims) = sum_batching_rule(self_, self_bdims, dim, keepdim, dtype);
```
How come `sum_batching_rule` doesn't just return a BatchedTensor from the getgo?

ezyang replied:
This is probably what you are referring to with 

> Batching rules and utility functions act on "unpacked BatchedTensors".
> This is to avoid tensor allocation overhead (one alternative is having
> these act on BatchedTensors). Unfortunately this leads to more verbose
> and confusing code. I've considered adding a special
> "BatchedTensorData" struct that only holds a Tensor and BatchDims and
> has helper methods defined on it but I'm not sure if that is better.

But I don't understand how passing things in expanded form helps avoid allocations.

ezyang replied:
Is the allocation avoidance specifically for the case when we mixed batched and non-batched tensors?

zou3519 replied:
I wanted the batching rules to look similar to their TorchScript variants. Right now I am envisioning that we have these batching rules in C++, and then a separate TorchScript set. The TorchScript batching rules should take in tensors and return tensors, because it would attempt to rewrite the graph using regular tensors.

However, that is not set in stone and I probably won't worry about it for ~2 months. I'm not sure if it is better to make the simplifying assumption that "batching rules should return BatchedTensors" and then worry about how TorchScript batching rules will look like in the future

zou3519 replied:
Yeah, the allocation avoidance advice is specifically for the case when we mix batched and non-batched -- wrapping the non-batched tensor in a BatchedTensorImpl would be unnecessary. But maybe that is premature optimization.

The main reason for the batching rules to look like this is because I wanted them to look similar to however we would add them in TorchScript in the future, but maybe this is also premature optimization

ezyang replied:
OK, definitely worth writing down in code what you are planning for in the future, no matter what.

The form of the TorchScript translations is an interesting problem. I actually think you should consider working out how to make TorchScript vmap rules accept batch tensors, and not the deconstructed tensors. I imagine how I would go about doing this in a good old fashioned compiler and the easiest way to eliminate vmap from a program is to simply inline all the implementations of vmapped operations into the use sites. To do this inlining, you have construct a semantic correspondence between the batched tensors (in the source program) with the batched tensors of the inputs to the batching rules (the function being inlined). 

So no matter what, you need a way to tell that the pair of Tensor and BDims corresponded to a batch tensor in the source program. So you could have some rule that constructs this correspondence... or you could have just passed them in as batched tensors to begin with!

What I am not so sure about is how convenient it will be to do this sort of thing in TorchScript as the compiler is setup today.

zou3519 replied:
> So no matter what, you need a way to tell that the pair of Tensor and BDims corresponded to a batch tensor in the source program. So you could have some rule that constructs this correspondence... or you could have just passed them in as batched tensors to begin with!

That's a good argument for passing the tensors in as BatchedTensors. I am not sure how convenient it is to do this in TorchScript either, but it is something I'll need to figure out down the line

ezyang commented on aten/src/ATen/BatchingRules.cpp:
```diff
@@ -0,0 +1,31 @@
+#include <ATen/BatchingRules.h>
+#include <ATen/ATen.h>
+
+namespace at {
```
I think what would be super useful is to include, at the top of this level (or some relevant file), some high level explanation about how to go about writing one of these batching rules. I bet some of this is in your design docs and just needs to be copy pasted here. Some questions an average user might ask:

* What does a batching rule function do?
* Can I reuse existing batching rule functions? When do I have to write a new one?
* Is there a recipe for writing a batching rule function?

ezyang replied:
Put differently, we should give users some abstract model to work with. The abstract model should be obviously correct but not implementable, and then can be used to determine if the implementable version. I'm personally partial to the "whenever you do a vmap, this causes a broadcast dimension to be added to every non-vmap'ed tensor in scope" explanation.

ezyang commented on aten/src/ATen/BatchingRules.cpp:
```diff
@@ -0,0 +1,31 @@
+  // One alternative way to do this is to keep them where they are and compute
+  // the required `dims` to reduce over. However, assuming that the batch
+  // dims are at front greatly simplifies the `dims` calculation and moving
+  // them there is relatively cheap.
```
Should this "move to the front" simply be backed in as an invariant to the batch dimensions, so you can just assume batch dimensions are in canonical form? What would go wrong otherwise?

ezyang replied:
Ah, I think I remember: the reason you don't do this is because you want to avoid repeatedly transposing the tensor (the way things are setup now, you only need to transpose once, right before use, rather than transposing on very nested vmap). If that's the case, it seems like it another way of explaining `BatchedTensorImpl` is that it records a "delayed" permutation of some sort. In other words, `moveBatchDimsToFront` isn't just a utility function: it is central to the *semantics* of batched tensors.

Although, it is not a complete semantics, since moveBatchDimsToFront really is only going to work for single tensor input operations, right? The multiargument ones are probably getting their own utilities?

zou3519 replied:
> Ah, I think I remember: the reason you don't do this is because you want to avoid repeatedly transposing the tensor (the way things are setup now, you only need to transpose once, right before use, rather than transposing on very nested vmap). 

That's part of the reason. Another reason is that there are some ops that perform better when the batch dimension is not at the front (only LSTMs), so the hypothesis is that in the future, we might want some batching rules to not permute dimensions if it impacts performance. 

Alternatively, we could just say this is an invariant for eager-mode vmap and look to the JIT to improve transposing.

> Although, it is not a complete semantics, since moveBatchDimsToFront really is only going to work for single tensor input operations, right? The multiargument ones are probably getting their own utilities?

Yup, the multiargument ones need their own utilities to align dimensions

zou3519 replied:
> That's part of the reason. Another reason is that there are some ops that perform better when the batch dimension is not at the front (only LSTMs), so the hypothesis is that in the future, we might want some batching rules to not permute dimensions if it impacts performance.

I guess it doesn't solve the problem if some (most) batching rules move dimensions to the front, but some batching rules don't. If we had a "time, batch, channel" tensor and a vmap over "mul" and then "lstm", "mul" would move the batch dimension to the front first and then the batch dimension would not be in the middle anymore for lstm

ezyang replied:
I mean, it doesn't *really* matter, because the physical layout is the same, and that's what counts, as far as efficiency goes. (Actually, now I wonder how vmap is going to interact with memory layout! If we transpose internal batch dimensions to the left, will they then durably become batch dimensions on the outside?)

ezyang replied:
It seems like it vmap is very related to memory layout, because preserving layout is essential to maintaining semantics for inner batch dimensions.

To be super concrete, let's imagine that we are doing a pointwise operation on a `(1,2,3)` size contiguous tensor, where the 2nd dimension is marked as batch dimension in vmap. At the end of the operation, we should end up with a `(1,2,3)` size contiguous tensor. However, if we transpose, and the pointwise operation doesn't preserve memory layout, you will end up with a physical layout of `(2,1,3)`. In the case of pointwise you can "work around" this by simply not transposing anything, but for example for convolution, you're not going to have a kernel that is going to be able to deal with an internal batch dimension, and so you are going to end up with the "wrong" layout in the end.

cc @VitalyFedyunin, what is the memory layout plan here?

zou3519 replied:
> preserving layout is essential to maintaining semantics for inner batch dimensions.

I'm not sure that it is important (or plausible) for vmap to maintain how tensors are strided for inner batch dimensions. The value-statement for vmap is that it makes batching convenient but won't be as fast as manual batching; if users want to fine-tune how the strides of their tensors are for performance reasons, then they're better off doing manual batching.

> but for example for convolution, you're not going to have a kernel that is going to be able to deal with an internal batch dimension, and so you are going to end up with the "wrong" layout in the end.

It looks like in order to support "maintaining strides semantics" for inner batch dimension, then we would need a kernel for each layout here... (or just a generic one that doesn't rely on contiguous or NHWC layout)

ezyang replied:
> I'm not sure that it is important (or plausible) for vmap to maintain how tensors are strided for inner batch dimensions. 

I agree. I think all I am asking for is a semantics. A user asks "what do the strides of vmap look like at the end" and we should be able to answer this question, in much the way we answer the question for most regular operators.

ezyang commented on aten/src/ATen/BatchingUtils.h:
```diff
@@ -0,0 +1,39 @@
+// For example:
+//   moveBatchDimsToFront([(lvl=1, dim=2), (lvl=3, dim=1)])
+// returns:
+//   [(lvl=1, dim=0), (lvl=3, dim=0)]
```
dim=0 in both cases?!

zou3519 replied:
Oops, copy-paste accident. Thanks for the catch!

ezyang commented:
The code looks correct. My concerns primarily concern around whether or not other developers will be able to understand how to write batching rules.

zou3519 committed:
Update on "Add "batching rule" for torch.sum(tensor, dims)"


On the way to moving vmap prototype into tree, this is the
implementation of the sum batching rule and some utility functions.
After implementing this and another batching rule (probably for
at::mul), we'll be ready to for the initial vmap Python
API.

Batching rules and utility functions act on "unpacked BatchedTensors".
This is to avoid tensor allocation overhead (one alternative is having
these act on BatchedTensors). Unfortunately this leads to more verbose
and confusing code. I've considered adding a special
"BatchedTensorData" struct that only holds a Tensor and BatchDims and
has helper methods defined on it but I'm not sure if that is better.

Re: testing: There's a bit of a chicken-and-egg problem here; I'd like to test
that the batching rules work for vmap, but the easiest way to do that is using
the (non-existent) vmap python API. This PR includes some minimal testing for
this batching rule.

Future PRs (for more context):
- [YOU ARE HERE] Add torch.sum(tensor, dims) batching rule + some helper functions
- Add torch.mul batching rule + some more helper functions
- Add vmap python API
- Add "slow" for-loop fallbacks for out-of-place functions via
dispatcher fallback mechanism.
- Add batching rules for "view" functions
- Add "slow" for-loop fallbacks for in-place functions
- And more

Test Plan:
- `./build/bin/vmap_test`

[ghstack-poisoned]

zou3519 commented:
I agree with the concerns that it's difficult to understand how to write batching rules. I added a large comment to describe the process, but I think it would be nice if I could simplify things more (I'm still thinking about whether or not it is OK to just accept and receive BatchedTensors in the batching rules)

zou3519 commented:
@ezyang I pondered over the concerns here and I am thinking of removing the restriction that "Batching Rules" must operate on "unpacked BatchedTensors". 

The original two reasons I wanted to have Batching Rules operate on "unpacked BatchedTensors" were:
1. Make batching rules in the JIT look pretty similar to their eager mode counterparts
2. Avoid unnecessary unwrapping and rewrapping of BatchedTensor

For (1), I think it's better to punt that to the future so we have something that looks sane for now rather than look weird due to murky requirements (I can revisit and refactor the API once we get closer to figuring out how vmap in the JIT should look like). For (2), we can still avoid unnecessary unwrap+wrap even if we operate on BatchedTensors, we just have to be careful when to unwrap and wrap in a batching rule.

In the new proposed scheme, a "Batching Rule" is the thing that gets registered as the operator. So each batching rule would have the same signature as the operator. 
```
m.impl_UNBOXED("sum.dim_IntList", sum_batching_rule);
```

The `sum_batching_rule` would operate on BatchedTensors, like the following:
```
Tensor sum_batching_rule(const Tensor& self, IntArrayRef dims, bool keepdim, c10::optional<ScalarType> dtype) {
    // moveBatchDimsToFront takes in either a Tensor or a BatchedTensor,
    // and always returns a regular Tensor (never a BatchedTensor!)
    Tensor self_ = moveBatchDimsToFront(self);
    new_dims = ... // add num_bdims to each of dims
    return makeBatched(self_.sum(new_bdims), ...)
}
```

`moveBatchDimsToFront` is the thing that is responsible for unwrapping a BatchedTensor and moving batch dimensions to the front. It only works for single-arg operators, but we'll have a multi-arg analogue (something like the following)

```
// Takes a list of Tensors. Each Tensor may be a BatchedTensor or a regular Tensor.
//
// materializeBatchDimsAtFront moves all the batch dimensions (if they exist)
// to the front of their respective tensors, ordered by level. Furthermore,
// it aligns all of the batch dimensions of the tensors, adding extra dimensions
// of size 1 if a tensor did not have that dimension. Finally, it returns non-batched
// tensors that can be directly operated on.
//
// The result is that each output Tensor has the same number of (public) batch
// dimensions that are aligned with respect to each tensor. We also return a
// bitset of which levels correspond to the (public) batch dimensions in the
// output.
std::pair<std::vector<Tensor>,std::bitset<64>>
materializeBatchDimsAtFront(TensorList tensors);
```

This should make it easier to write batching rules so that the user does not have to juggle "unpacked BatchedTensors". Does this sound reasonable?

ezyang commented:
Yes, batching rules as registered operators makes a good deal of sense to me.

zou3519 committed:
Update on "Add "batching rule" for torch.sum(tensor, dims)"


On the way to moving vmap prototype into tree, this is the
implementation of the sum batching rule and some utility functions.
After implementing this and another batching rule (probably for
at::mul), we'll be ready to for the initial vmap Python
API.

Re: testing: There's a bit of a chicken-and-egg problem here; I'd like to test
that the batching rules work for vmap, but the easiest way to do that is using
the (non-existent) vmap python API. This PR includes some minimal testing for
this batching rule.

Future PRs (for more context):
- [YOU ARE HERE] Add torch.sum(tensor, dims) batching rule + some helper functions
- Add torch.mul batching rule + some more helper functions
- Add vmap python API
- Add "slow" for-loop fallbacks for out-of-place functions via
dispatcher fallback mechanism.
- Add batching rules for "view" functions
- Add "slow" for-loop fallbacks for in-place functions
- And more

Test Plan:
- `./build/bin/vmap_test`

[ghstack-poisoned]

ezyang commented on aten/src/ATen/BatchingUtils.h:
```diff
@@ -0,0 +1,42 @@
+// Produces a new vector of the same size as `arr` where each element of is
+// created by calling `func` on  each element of `arr` in order.
+// Equivalent to `output = map(func, arr)` in Python.
+std::vector<int64_t> transform(IntArrayRef arr, std::function<int64_t(int64_t)> func);
```
This is a pretty generic name that's been put in the top level `at` namespace. Try making it more descriptive or putting it in a subnamespace.

zou3519 replied:
Good point. I made it more descriptive. In the long term I want to move most of the BatchedTensor things into its own subnamespace to avoid crowding the at:: namespace

zou3519 committed:
Update on "Add "batching rule" for torch.sum(tensor, dims)"


On the way to moving vmap prototype into tree, this is the
implementation of the sum batching rule and some utility functions.
After implementing this and another batching rule (probably for
at::mul), we'll be ready to for the initial vmap Python
API.

Re: testing: There's a bit of a chicken-and-egg problem here; I'd like to test
that the batching rules work for vmap, but the easiest way to do that is using
the (non-existent) vmap python API. This PR includes some minimal testing for
this batching rule.

Future PRs (for more context):
- [YOU ARE HERE] Add torch.sum(tensor, dims) batching rule + some helper functions
- Add torch.mul batching rule + some more helper functions
- Add vmap python API
- Add "slow" for-loop fallbacks for out-of-place functions via
dispatcher fallback mechanism.
- Add batching rules for "view" functions
- Add "slow" for-loop fallbacks for in-place functions
- And more

Test Plan:
- `./build/bin/vmap_test`

[ghstack-poisoned]

ezyang commented on aten/src/ATen/BatchingUtils.h:
```diff
@@ -0,0 +1,42 @@
+// Produces a new vector of the same size as `arr` where each element of is
+// created by calling `func` on  each element of `arr` in order.
+// Equivalent to `output = map(func, arr)` in Python.
+std::vector<int64_t> transformIntVector(IntArrayRef arr, std::function<int64_t(int64_t)> func);
```
Use fmap from `ATen/core/functional.h`?

```
// Overload for explicit function and ArrayRef
template<class F, class T>
inline auto fmap(const T& inputs, const F& fn) -> std::vector<decltype(fn(*inputs.begin()))> {
  std::vector<decltype(fn(*inputs.begin()))> r;
  r.reserve(inputs.size());
  for(const auto & input : inputs)
    r.push_back(fn(input));
  return r;
}
```

ezyang commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -0,0 +1,87 @@
+  auto actual_dims = transformIntVector(
+      dims,
+      [&](int64_t dim) {
+        return maybe_wrap_dim(dim, self_dim) + num_bdims;
```
I don't think wrapping on `self_dim` is correct. When we vmap over batch dimensions, the logical dimension should decrease. That means self dim should decrease here as well.

zou3519 replied:
`self_dim` is the logical dimensionality, `dims` is a logical dimension (passed in by the user), and `actual_dims` represents physical dimensions, so I think this is correct. But I do see the confusion

cpuhrsch replied:
Might it be worth to create a special-case function for this? I'd imagine this will be very common and potentially confusing. I'd also add a sentence about this in above documentation as the sort-of work that might be unexpected.

ezyang commented:
I still feel like it is not all that clear how to go about writing batching rules. One way you might be able to get more signal one way or another is to ask someone else to look over the sum rule and see if they can figure out what it is doing.

The distinction that I feel like I'm looking for is a more clear distinction between "logical" and "physical" views on tensors. For example, a tensor might be (3,2,4), but if it is batched on dim 0 and dim 2, then its logical view is (2). So for example, I have a batched tensor, and I want to know what its logical dimensionality is. There should be some way to get this directly. (c.f. https://github.com/pytorch/pytorch/pull/38723/files#r433280455 ). Or, I have some logical dimension references (e.g. dims) and I need to map them into physical dimensions so I can call the underlying call. I shouldn't have to do this manually with transformIntVector.

cpuhrsch commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -0,0 +1,87 @@
+
+// Note: [How to write batching rules]
+//
+// Whenever you do a vmap, on a vmap'ed tensor, the dimension that is being vmap'ed
```
This paragraph sounds a bit strange. Maybe this is better (not a native speaker):

"""For each level of vmap the specified dimension is recorded. Once the innermost vmap was reached the recorded batch dimensions act as broadcasting dimensions for the given operation."""

cpuhrsch commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -0,0 +1,87 @@
+// `
+// Most batching rules will look like the following:
+// 1. Move all batch dims to the front of the input tensors
+// 2. If there are multiple tensors, align their batch dimensions by level
```
I'd expand a bit on this sentence, since level wasn't introduced yet.

zou3519 replied:
Level is introduced in BatchedTensorImpl.h.https://github.com/pytorch/pytorch/blob/baf6ed0238df7a8f30655df5b3b265e77787e422/aten/src/ATen/BatchedTensorImpl.h#L19-L22

 But I can make that a note and refer back to it

cpuhrsch commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -0,0 +1,87 @@
+// 1. Move all batch dims to the front of the input tensors
+// 2. If there are multiple tensors, align their batch dimensions by level
+// 3. Once the tensors are aligned, figure out how to call the at:: op that
+//    corresponds to the operator with additional batch dims
```
You mean "the operator with broadcasting support" or such?

zou3519 replied:
The operator doesn't have to have broadcasting support. It could be an operator that accepts a single batch dimension (like torch.bmm), or an operator that accepts arbitrary broadcasting dimensions (like torch.mul), or an operator that accepts arbitrary broadcasting batch dimensions (like torch.solve)

cpuhrsch commented on aten/src/ATen/BatchingUtils.h:
```diff
@@ -0,0 +1,42 @@
+// Equivalent to `output = map(func, arr)` in Python.
+std::vector<int64_t> transformIntVector(IntArrayRef arr, std::function<int64_t(int64_t)> func);
+
+// Takes a BatchedTensor or a regular Tensor and permutes all of the batch dims of
```
The "or a regular Tensor" makes it sound like there is value in this function if it's not in fact a BatchedTensor. I'd write out the case for BatchedTensor and then also say that this acts as an identity operation (for convenience) if the given Tensor is not a BatchedTensor.

zou3519 commented:
The stack at https://github.com/pytorch/pytorch/pull/39580 supersedes this PR.

From some offline discussion, we have come up with a new API for batching rules that adds more structure to the process of writing one. In particular, it abstracts away logical and physical views on tensors and offers a mapping mechanism between logical and physical views and also logical and physical arguments (like dimension indices, shapes, and more).
